### PR TITLE
Migration to netstandard 1.4

### PR DESCRIPTION
**Pull request by mayankbansal018**
1. Migrating CoreUtilites to netstandard1.4
2. Adding a platform abstraction layer, to implement platform specific requirements

### CODE REVIEW

File: scripts/build.ps1 [hunk 1]
```diff
@@ -252,6 +252,22 @@ function Publish-Package
     Write-Log "Publish-Package: Complete. {$(Get-ElapsedTime($timer))}"
 }
 
+function Patch-Publish-Package
```
Comment (codito) [inline scripts/build.ps1]: Publish-PlatfromAbstractions, call from within Publish-Package.

File: scripts/build.ps1 [hunk 2]
```diff
@@ -252,6 +252,22 @@ function Publish-Package
...
+{
+    $timer = Start-Timer
+    Write-Log "Publish-Package: Started."
+    $dotnetExe = Get-DotNetPath
+    $fullCLRPackageDir = Get-FullCLRPackageDirectory
+    $coreCLRPackageDir = Get-CoreCLRPackageDirectory
+	
+	$platformAbstraction = Join-Path $env:TP_ROOT_DIR "src\Microsoft.TestPlatform.PlatformAbstractions\bin\$TPB_Configuration"
```
Comment (codito) [inline scripts/build.ps1]: nit:indent

File: src/Microsoft.TestPlatform.CrossPlatEngine/Microsoft.TestPlatform.CrossPlatEngine.csproj [hunk 3]
```diff
@@ -30,6 +30,9 @@
     <PackageReference Include="System.Diagnostics.Process">
       <Version>4.3.0</Version>
     </PackageReference>
+    <PackageReference Include="System.Threading.Thread">
```
Comment (codito) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Microsoft.TestPlatform.CrossPlatEngine.csproj]: Was this dependency getting implicitly pulled in earlier?

File: src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/Tracing/IEqtTrace.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.Diagnostics;
+
+namespace Microsoft.VisualStudio.TestPlatform.ObjectModel
+{
+    //
```
Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/Tracing/IEqtTrace.cs]: nit: `///` for summaries.

File: src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/Tracing/IEqtTrace.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    // Summary:
+    //     Specifies what messages to output for the System.Diagnostics.Debug, System.Diagnostics.Trace
+    //     and System.Diagnostics.TraceSwitch classes.
+    public enum CustomTraceLevel
```
Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/Tracing/IEqtTrace.cs]: Should we call this `TestPlatformTraceLevel`?

**Comment by codito** [comment]
Remove `ProjectGuid` from PlatformAbstractions csproj.

File: src/Microsoft.TestPlatform.PlatformAbstractions/net46/Tracing/RollingFileTraceListener.cs [hunk 6]
```diff
@@ -37,7 +34,7 @@ public class RollingFileTraceListener : TextWriterTraceListener
         {
             if (string.IsNullOrWhiteSpace(fileName))
             {
-                throw new ArgumentException(Resources.CannotBeNullOrEmpty, nameof(fileName));
+                throw new ArgumentException(nameof(fileName));
```
Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/net46/Tracing/RollingFileTraceListener.cs]: Is the string `Resources.CannotBeNullOrEmpty` redundant?

File: src/Microsoft.TestPlatform.PlatformAbstractions/netcore/Tracing/EqtTrace.cs [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+namespace Microsoft.VisualStudio.TestPlatform.ObjectModel
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Globalization;
+    using System.IO;
+    using System.Threading;
+
+    /// <summary>
```
Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/netcore/Tracing/EqtTrace.cs]: We can move documentation for public members to `IPlatformEqtTrace` and use `/// <inheritdoc/>` to reduce duplication.

File: src/TestPlatform.TranslationLayer.nuspec [hunk 8]
```diff
@@ -49,12 +49,14 @@
     <file src="net46\$Runtime$\Microsoft.TestPlatform.VsTestConsole.TranslationLayer.dll" target="lib\net46\" />
     <file src="net46\$Runtime$\Microsoft.TestPlatform.CommunicationUtilities.dll" target="lib\net46\" />
     <file src="net46\$Runtime$\Microsoft.TestPlatform.CoreUtilities.dll" target="lib\net46\" />    
+	<file src="net46\$Runtime$\Microsoft.TestPlatform.PlatformAbstractions.dll" target="lib\net46\" />
```
Comment (codito) [inline src/TestPlatform.TranslationLayer.nuspec]: nit: indent. Same for below.

File: src/TestPlatform.ObjectModel.nuspec [hunk 9]
```diff
@@ -49,9 +49,11 @@
   <files>	
     <file src="net46\$Runtime$\Microsoft.VisualStudio.TestPlatform.ObjectModel.dll" target="lib\net46\" />
     <file src="net46\$Runtime$\Microsoft.TestPlatform.CoreUtilities.dll" target="lib\net46\" />
+	<file src="net46\$Runtime$\Microsoft.TestPlatform.PlatformAbstractions.dll" target="lib\net46\" />
```
Comment (codito) [inline src/TestPlatform.ObjectModel.nuspec]: nit: indent. Same for below.

File: src/Microsoft.TestPlatform.PlatformAbstractions/netcore/Tracing/EqtTrace.cs [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+    ///     - Uses custom switch <c>EqtTraceLevel</c> from .config file.
+    ///     - By default tracing if OFF.
+    ///     - Our build environment always sets the /d:TRACE so this class is always enabled,
+    ///       the Debug class is enabled only in debug builds (/d:DEBUG).
+    ///     - We ignore exceptions thrown by underlying TraceSwitch (e.g. due to config file error).
+    ///       We log ignored exceptions to system Application log.
+    ///       We pass through exceptions thrown due to incorrect arguments to <c>EqtTrace</c> methods.
+    /// Usage: <c>EqtTrace.Info("Here's how to trace info");</c>
+    /// </summary>
+    public class PlatformEqtTrace : IPlatformEqtTrace
```
Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/netcore/Tracing/EqtTrace.cs]: What is the test strategy for this class?

**Review by codito — COMMENTED** [review]
Looks good with few comments.

File: src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/Tracing/IEqtTrace.cs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+        void InitializeVerboseTrace(string customLogFile);
+
+        bool ShouldTrace(CustomTraceLevel traceLevel);
+
+        string GetLogFile();
+
+        void SetTraceLevel(CustomTraceLevel value);
+
+#if NET46
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/Tracing/IEqtTrace.cs]: :( Isn't this smell in an interface?

File: src/Microsoft.TestPlatform.PlatformAbstractions/netstandard1.0/Tracing/EqtTrace.cs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    ///     - Our build environment always sets the /d:TRACE so this class is always enabled,
+    ///       the Debug class is enabled only in debug builds (/d:DEBUG).
+    ///     - We ignore exceptions thrown by underlying TraceSwitch (e.g. due to config file error).
+    ///       We log ignored exceptions to system Application log.
+    ///       We pass through exceptions thrown due to incorrect arguments to <c>EqtTrace</c> methods.
+    /// Usage: <c>EqtTrace.Info("Here's how to trace info");</c>
+    /// </summary>
+    public class PlatformEqtTrace : IPlatformEqtTrace
+    {
+        public void WriteLine(CustomTraceLevel level, string message)
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.PlatformAbstractions/netstandard1.0/Tracing/EqtTrace.cs]: We are planning to fill this in later? Looks like there might be common code with desktop for this.

File: src/Microsoft.TestPlatform.CoreUtilities/Tracing/EqtTrace.cs [hunk 13]
```diff
@@ -29,106 +23,34 @@ namespace Microsoft.VisualStudio.TestPlatform.ObjectModel
...
+        public static string LogFile
         {
             get
             {
-                return SourceTraceLevelsMap[Source.Switch.Level];
+                return traceImpl.GetLogFile();
             }
+        }
 
+        public static CustomTraceLevel TraceLevel
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.CoreUtilities/Tracing/EqtTrace.cs]: Isn't this a breaking change?

File: src/Microsoft.TestPlatform.PlatformAbstractions/net46/Tracing/RollingFileTraceListener.cs [hunk 14]
```diff
@@ -37,7 +34,7 @@ public class RollingFileTraceListener : TextWriterTraceListener
         {
             if (string.IsNullOrWhiteSpace(fileName))
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.PlatformAbstractions/net46/Tracing/RollingFileTraceListener.cs]: Suggestion : Can use ValidateArg.NotNullOrEmpty here.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Microsoft.TestPlatform.CrossPlatEngine.csproj (hunk 3 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Microsoft.TestPlatform.CrossPlatEngine.csproj]: yes It was coming from CoreUtilities.csproj

File: src/Microsoft.TestPlatform.PlatformAbstractions/net46/Tracing/RollingFileTraceListener.cs (hunk 14 above)
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.PlatformAbstractions/net46/Tracing/RollingFileTraceListener.cs]: Can't use, it will create circular dependency

**Review by codito — APPROVED** [review]
With few suggestions

**Comment by codito** [comment]
@mayankbansal018 please ensure `sign.proj` includes the new assemblies. It will break TestPlatform.CI.Real otherwise.

**Pull request merged**